File: MaximalRectangle.java
class Solution {
    int largestRectangleArea(int[] heights) {
    	int[] stack = new int[heights.length + 1];
    	stack[0] = -1;
    	int top = 0;
    	int maxArea = 0;
    	for(int i = 0;i < heights.length;i++){
    		if(top <= 0 || heights[i] > heights[stack[top]]){
    			stack[++top] = i;
    		}else{
    			while(top > 0 && heights[i] <= heights[stack[top]]){
        			int tmp = heights[stack[top]] * (i - stack[top - 1] - 1);
        			maxArea = tmp > maxArea ? tmp : maxArea;
        			top--;	
    			}
    			stack[++top] = i;
    		}
    	}
    	for(int i = top;i > 0;i--){
    		int tmp = heights[stack[i]] * (stack[top] - stack[i - 1]);
    		maxArea = tmp > maxArea ? tmp : maxArea;
    	}
    	return maxArea;
    }
	
    public int maximalRectangle(char[][] matrix) {
    	int maxArea = 0;
    	if(matrix.length <= 0) return 0;
    	int[] val = new int[matrix[0].length];
        for(int i = 0;i < matrix.length;i++){
        	for(int j = 0;j < val.length;j++){
        		val[j] = (val[j] + 1) * (matrix[i][j] - '0');
        	}
        	int r = largestRectangleArea(val);
        	maxArea = r > maxArea ? r : maxArea;
        }
        return maxArea;
    }
}

File: MaximalRectangle.cpp
#include <vector>
#include <cmath>

using namespace std;

class Solution {
public:
    int maximalRectangle(vector<vector<char>>& matrix) {
        const int m = matrix.size();
        if (m == 0)
            return 0;
        const int n = matrix[0].size();
        int left[n] = {0};
        int right[n] = {n};
        int height[n] = {0};

        int cur_left, cur_right;

        int maxarea = 0;
        for (int i = 0; i < m; i++)
        {
            cur_left = 0;
            cur_right = n;
            // update height
            for (int j = 0; j < n; j++)
                if (matrix[i][j] == '1')
                    height[j]++;
                else
                    height[j] = 0;
            // update left
            for (int j = 0; j < n; j++)
                if (matrix[i][j] == '1')
                {
                    if (left[j] < cur_left)
                        left[j] = cur_left;
                }
                else
                {
                    left[j] = 0;
                    cur_left = j + 1;
                }
            // update right
            for (int j = n - 1; j >= 0; j--)
                if (matrix[i][j] == '1')
                {
                    if (right[j] > cur_right)
                        right[j] = cur_right;
                }
                else
                {
                    right[j] = n;
                    cur_right = j;
                }
            // update area
            for (int j = 0; j < n; j++)
                maxarea = max(maxarea, abs(right[j] - left[j]) * height[j]);
        }
        return maxarea;
    }
};